File: src/shared/clock.test.ts
import { describe, expect, it } from "vitest";
import {
  blendOffset,
  calculateClockSample,
  calculateJitterMs,
  CLOCK_OFFSET_JUMP_MS,
  delayUntilServerTime,
  isClockConverged,
  summarizeClock
} from "./clock.js";

describe("clock sync", () => {
  it("calculates offset and round trip time from NTP-style timestamps", () => {
    const sample = calculateClockSample(1000, 1100, 1055, 1060);

    expect(sample.rttMs).toBe(95);
    expect(sample.offsetMs).toBe(7.5);
  });

  it("takes the offset from the single lowest-rtt sample", () => {
    const summary = summarizeClock([
      { rttMs: 100, offsetMs: 30 },
      { rttMs: 20, offsetMs: 10 },
      { rttMs: 25, offsetMs: 12 },
      { rttMs: 500, offsetMs: 200 }
    ]);

    // RTT is still the median of the best samples for the quality gauge...
    expect(summary.rttMs).toBe(62.5);
    // ...but the offset is the cleanest round trip (rtt 20 -> offset 10).
    expect(summary.offsetMs).toBe(10);
  });

  it("returns a zeroed summary for no samples", () => {
    expect(summarizeClock([])).toEqual({ rttMs: 0, offsetMs: 0 });
  });

  it("converts scheduled server time to a local timeout delay", () => {
    expect(delayUntilServerTime(10_000, 8_000, 1_000)).toBe(1_000);
    expect(delayUntilServerTime(10_000, 12_000, 0)).toBe(0);
  });

  it("calculates jitter from offset variation", () => {
    expect(calculateJitterMs([
      { rttMs: 20, offsetMs: 10 },
      { rttMs: 21, offsetMs: 12 },
      { rttMs: 19, offsetMs: 15 }
    ])).toBe(2);
  });

  describe("blendOffset", () => {
    it("adopts the first measurement outright", () => {
      expect(blendOffset(undefined, 2000)).toBe(2000);
    });

    it("eases small changes in to damp jitter", () => {
      // 100 -> 120 with 0.3 smoothing lands at 106, not the full 120.
      expect(blendOffset(100, 120, 0.3)).toBeCloseTo(106);
    });

    it("adopts a large jump immediately (real clock step)", () => {
      const next = 100 + CLOCK_OFFSET_JUMP_MS + 50;
      expect(blendOffset(100, next)).toBe(next);
    });
  });

  describe("isClockConverged", () => {
    it("requires enough samples and acceptable jitter", () => {
      expect(isClockConverged(5, 10)).toBe(true);
      expect(isClockConverged(2, 10)).toBe(false);
      expect(isClockConverged(10, 40)).toBe(false);
    });
  });
});

File: android/app/src/test/java/com/bandcue/songsterr/ClockTest.kt
package com.bandcue.songsterr

import org.junit.Assert.assertEquals
import org.junit.Test

class ClockTest {
    @Test
    fun calculatesClockSampleLikeWebClient() {
        val sample = calculateClockSample(
            clientSentAt = 1_000,
            clientReceivedAt = 1_120,
            serverReceivedAt = 2_040,
            serverSentAt = 2_050
        )

        assertEquals(110.0, sample.rttMs, 0.01)
        assertEquals(985.0, sample.offsetMs, 0.01)
    }

    @Test
    fun summarizesBestRttSamples() {
        val summary = summarizeClock(
            listOf(
                ClockSample(250.0, 12.0),
                ClockSample(20.0, 4.0),
                ClockSample(10.0, 2.0),
                ClockSample(30.0, 6.0),
                ClockSample(40.0, 8.0),
                ClockSample(50.0, 10.0)
            )
        )

        // RTT is still the median of the best samples for the quality gauge...
        assertEquals(30.0, summary.rttMs, 0.01)
        // ...but the offset comes from the single lowest-RTT sample (rtt 10 -> 2).
        assertEquals(2.0, summary.offsetMs, 0.01)
    }

    @Test
    fun blendOffsetAdoptsFirstAndLargeJumps() {
        assertEquals(2_000.0, blendOffset(null, 2_000.0), 0.01)
        val jump = 100.0 + CLOCK_OFFSET_JUMP_MS + 50.0
        assertEquals(jump, blendOffset(100.0, jump), 0.01)
    }

    @Test
    fun blendOffsetEasesSmallChanges() {
        assertEquals(106.0, blendOffset(100.0, 120.0, 0.3), 0.01)
    }

    @Test
    fun reportsConvergenceFromSamplesAndJitter() {
        assertEquals(true, isClockConverged(5, 10.0))
        assertEquals(false, isClockConverged(2, 10.0))
        assertEquals(false, isClockConverged(10, 40.0))
    }

    @Test
    fun appliesServerOffsetWhenCalculatingDelay() {
        val delay = delayUntilServerTime(
            scheduledServerTime = 2_000,
            localNow = 1_000,
            serverOffsetMs = 250.0
        )

        assertEquals(750, delay)
    }
}
